### PR TITLE
Keystore

### PR DESCRIPTION
**Pull request by humanjas**
#160, #161 

# Introduction
The private and public keys are now saved using the AES-CTR protocol, using the hex representation of the public key of the user as password. Android Keystore does not support ED25519 protocol, or in other words the Edward Elliptic Curve.

# Purpose
This way, the files containing the public/private keys, cannot be read out directly.

# Details
none

# Future Applicability
The public key as password, could be changed to something else.

### CODE REVIEW

**Comment by humanjas** [comment]
Pulled from the feature/travisTimeoutFix into this branch to increase the chance that the travis build succeeds.

File: app/src/test/java/nl/tudelft/b_b_w/controller/CryptoControllerTest.java [hunk 1]
```diff
@@ -7,20 +7,18 @@
...
-import nl.tudelft.b_b_w.model.User;
-
 import static org.junit.Assert.assertEquals;
 
 /**
  * Test class for CryptoController
  */
 public class CryptoControllerTest {
 
+    private static final byte[] TEST_SEED = Utils.hexToBytes("0000000000000000000000000000000000000000000000000000000000000000");
```
Comment (ashay123) [inline app/src/test/java/nl/tudelft/b_b_w/controller/CryptoControllerTest.java]: Is this really necessary? I think this is default

Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/CryptoControllerTest.java]: Changed

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java [hunk 2]
```diff
@@ -1,32 +1,38 @@
...
 public class CryptoController {
 
+    private static final String ENCRYPTION_EXTENSION = ".enc";
     /**
      * Class attributes
-     *
-     * @param textEncryptor The encryptor which uses AES Encryption in CBC mode with a maximum
-     * permitted key length of 256bit.
+     * <p>
+     * textEncryptor is the text encryptor uses AES encryption in CBC mode
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: I cannot understand this sentence.

File: .travis.yml [hunk 3]
```diff
@@ -107,4 +107,4 @@ after_script:
   - pandoc builder/build/reports/tests/index.html -t plain | sed -n '/^Failed tests/,/default-package/p'
 
 after_success:
-  - travis_wait 60 ./gradlew cobertura coveralls
\ No newline at end of file
+  - ./gradlew coveralls
```
Comment (ymte) [inline .travis.yml]: Why is this modified here?

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java [hunk 4]
```diff
@@ -58,4 +64,37 @@ public final String decryptString(String data) {
             throw new RuntimeException("Error - Could not decode string: " + e);
         }
     }
+
+    /**
+     * encryptFile method
+     *
+     * @param path given filepath to encrypt
+     */
+    @SuppressWarnings("ResultOfMethodCallIgnored")
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: Why is this warning suppressed?

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java [hunk 5]
```diff
@@ -58,4 +64,37 @@ public final String decryptString(String data) {
             throw new RuntimeException("Error - Could not decode string: " + e);
         }
     }
+
+    /**
+     * encryptFile method
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: This javadoc is useless.

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java [hunk 6]
```diff
@@ -58,4 +64,37 @@ public final String decryptString(String data) {
...
+     *
+     * @param path given filepath to encrypt
+     */
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    final void encryptFile(String path) {
+        try {
+            fileEncryptor.encrypt(new File(path + KeyWriter.TEMP_EXTENSION),
+                    new File(path + ENCRYPTION_EXTENSION));
+            new File(path + KeyWriter.TEMP_EXTENSION).delete();
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: Why not reuse the same File object?

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java [hunk 7]
```diff
@@ -58,4 +64,37 @@ public final String decryptString(String data) {
...
+    /**
+     * decryptFile method
+     *
+     * @param path given filepath to decrypt
+     */
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    final void decryptFile(String path) {
+        try {
+            fileEncryptor.decrypt(new File(path + ENCRYPTION_EXTENSION), new File(path));
+            new File(path + ENCRYPTION_EXTENSION).delete();
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: Same here.

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java [hunk 8]
```diff
@@ -58,4 +64,37 @@ public final String decryptString(String data) {
...
+            fileEncryptor.encrypt(new File(path + KeyWriter.TEMP_EXTENSION),
+                    new File(path + ENCRYPTION_EXTENSION));
+            new File(path + KeyWriter.TEMP_EXTENSION).delete();
+        } catch (GeneralSecurityException | IOException e) {
+            throw new RuntimeException("Error - Could not encrypt file: " + path);
+        }
+    }
+
+    /**
+     * decryptFile method
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: Same here.

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java [hunk 9]
```diff
@@ -58,4 +64,37 @@ public final String decryptString(String data) {
...
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    final void decryptFile(String path) {
+        try {
+            fileEncryptor.decrypt(new File(path + ENCRYPTION_EXTENSION), new File(path));
+            new File(path + ENCRYPTION_EXTENSION).delete();
+        } catch (GeneralSecurityException | IOException e) {
+            throw new RuntimeException("Error - Could not decrypt file: " + path);
+        }
+    }
+
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: Extra whitelines

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java [hunk 10]
```diff
@@ -49,7 +49,8 @@ private static void initialize(String filePath) throws IOException {
      * @param path given filepath
      * @return byte array containing the read key
      */
-    public static byte[] readKey(String path) throws IOException {
+    public static byte[] readKey(String path, String password) throws IOException {
+        decryptFile(path, password);
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Where is this stored? Why is the returned password not immediately returned?

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java [hunk 11]
```diff
@@ -69,9 +70,9 @@ private static void initialize(String filePath) throws IOException {
      *
      * @return private key
      */
-    public static EdDSAPrivateKey readPrivateKey() throws InvalidKeySpecException, IOException {
+    public static EdDSAPrivateKey readPrivateKey(String password) throws InvalidKeySpecException, IOException {
         final String privateKeyPath = "private.key";
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Hardcoded path file

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java [hunk 12]
```diff
@@ -81,20 +82,20 @@ public static EdDSAPrivateKey readPrivateKey() throws InvalidKeySpecException, I
...
+    public static EdDSAPublicKey readPublicKey(String password) throws InvalidKeySpecException, IOException {
         final String publicKeyPath = "public.key";
-        final byte[] encodedPublicKey = readKey(publicKeyPath);
+        final byte[] encodedPublicKey = readKey(publicKeyPath, password);
         return convertToPublicKey(encodedPublicKey);
     }
 
     /**
-     * readPublicKey method
+     * StringPublicKey method
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Useless javadoc line

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java [hunk 13]
```diff
@@ -81,20 +82,20 @@ public static EdDSAPrivateKey readPrivateKey() throws InvalidKeySpecException, I
...
      * Reads the public key from the given string
      *
      * @param encodedPublicKey given hex representation of the byte array of the public key
      * @return public key
      */
-    public static EdDSAPublicKey readPublicKey(String encodedPublicKey) throws InvalidKeySpecException {
+    public static EdDSAPublicKey StringToPublicKey(String encodedPublicKey) throws InvalidKeySpecException {
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Method should start with a lower case letter

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java [hunk 14]
```diff
@@ -123,4 +124,16 @@ private static EdDSAPublicKey convertToPublicKey(byte[] encodedPublicKey) throws
         X509EncodedKeySpec encoded = new X509EncodedKeySpec(encodedPublicKey);
         return new EdDSAPublicKey(encoded);
     }
+
+    /**
+     * decryptFile method
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: What does the "decryptFile method" line add?

File: app/src/test/java/nl/tudelft/bbw/controller/KeyStorageUnitTest.java [hunk 15]
```diff
@@ -84,26 +77,21 @@ public void testReadKeyException() throws IOException {
     @SuppressWarnings("ResultOfMethodCallIgnored")
     public void testReadKeyEmpty() throws IOException {
         final String empty = "empty";
-        exception.expect(IOException.class);
-        exception.expectMessage("File is empty: " + empty);
+        exception.expect(RuntimeException.class);
```
Comment (ymte) [inline app/src/test/java/nl/tudelft/bbw/controller/KeyStorageUnitTest.java]: Please do not create new empty files continually without cleaning them up after

File: .travis.yml (hunk 3 above)
Comment (humanjas) [inline .travis.yml]: As you can see in the commit, the cobertura command is already called, when the check command has been completed.

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 2 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: Changed

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 5 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: Changed

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 4 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: You are deleting a temporary file, which is needed for the encryption method, since you cannot overwrite the same file.

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 6 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: changed

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 8 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: changed

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 9 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: changed

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 10 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: This is stored in the path that you give, when you encrypt the file. There is no password returned.

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 11 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: That is because there is only a single user that is using this app. And this user can only have one private key. So there is no reason to store multiple private keys on a single device for this app.

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 12 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Since this is the case in a lot of classes and functions in our code, and we have to work consistently, this should not be changed or should be addressed in another pull request.

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 13 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: changed

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 14 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: It makes sure you do not write the same code multiple times and it decreases the amount of lines per method.

File: app/src/test/java/nl/tudelft/bbw/controller/KeyStorageUnitTest.java (hunk 15 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/bbw/controller/KeyStorageUnitTest.java]: Look at line 105 in this class.

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 5 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: I have not seen it anywhere in the code I work with. Consistency is not a reason for bad comments.

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 4 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: Why store it in a temporary file?

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 11 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Yes, but hardcoded strings should be above in `public static final` constants

File: app/src/test/java/nl/tudelft/bbw/controller/KeyStorageUnitTest.java (hunk 15 above)
Comment (ymte) [inline app/src/test/java/nl/tudelft/bbw/controller/KeyStorageUnitTest.java]: ok, clear

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 14 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: I mean the non-descriptive comment. Bad comments are worse than failed comments

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 12 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: That other classes are badly documented should not be a reason for this class to be badly documented as well.

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 10 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: But why is the temporarily decrypted key stored in a file?

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 5 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: It is not per definition bad and thus consistency is very important as well. Next to that, if you have a look at : https://github.com/deepweeb/B-B-W/tree/master/app/src/main/java/nl/tudelft/b_b_w, almost all controllers have them and almost all models too.

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 10 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Because you cannot overwrite a file, while you are reading it and the library does that, if it encrypts a file.

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 12 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Look at the other comment

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 5 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: I can see them now, and would like to refactor that at well. That code was not properly reviewed by me; I see no reason to repeat the method name appended with 'method' in the first line of the javadoc. Consistency is thus not important.

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 11 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: changed it

File: app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java (hunk 5 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/CryptoController.java]: According to our SEM TA, consistency of our whole code style is really important.

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 14 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Look at the comment above

File: app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java (hunk 10 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/KeyReader.java]: Ok, clear
